### PR TITLE
fixing all lessons (except lesson_3) in tutorial/4_imp++/

### PR DESCRIPTION
**Pull request by yilongli**
**UPDATE**
After fixing the long-lasting issue of mysterious cell order in the output, I changed the goal of this pull request to fixing all lessons of `imp++` except lesson_3, which requires the functionality of `superheat` and `supercool` in the Java backend.

**ORIGINAL DESCRIPTION**
After fixing a small issue in the pretty printer, `lesson_1` && `lesson_2` of imp++ are fine. I took the liberty to replace the *.out files with the generated outputs. I have checked that they should be correct, but please double check anyway.

I have not added them into regression tests, but I will do that once all lessons of imp++ are cleared.

@grosu @dwightguth please review

### CODE REVIEW

**Comment by grosu** [comment]
I just tried it and still gives some errors:

```
fslwork$ pwd
/home/grosu/github.grosu/k/tutorial/1_k/4_imp++/lesson_1
fslwork$ ktest tests/config.xml
Running initial scripts...
SUCCESS
Kompile the language definitions...(2 in total)
SUCCESS
Generate PDF files...(2 in total)
SUCCESS
Running /home/grosu/github.grosu/k/tutorial/1_k/4_imp++/lesson_1/imp.k programs... (8 in total, 0 with input, 8 with output, 0 with error)
Running /home/grosu/github.grosu/k/tutorial/1_k/4_imp++/lesson_1/imp.k programs... (8 in total, 0 with input, 8 with output, 0 with error)
ERROR: ['/home/grosu/github.grosu/k/bin/krun' '/home/grosu/github.grosu/k/tutorial/1_k/4_imp++/lesson_1/tests/io.imp' '--directory' 'imp-java_kompiled'] output doesn't match with expected output (time: 2111 ms)
ERROR: ['/home/grosu/github.grosu/k/bin/krun' '/home/grosu/github.grosu/k/tutorial/1_k/4_imp++/lesson_1/tests/spawn.imp' '--directory' 'imp-java_kompiled'] output doesn't match with expected output (time: 2252 ms)
ERROR: ['/home/grosu/github.grosu/k/bin/krun' '/home/grosu/github.grosu/k/tutorial/1_k/4_imp++/lesson_1/tests/sum-io.imp' '--directory' 'imp-java_kompiled'] output doesn't match with expected output (time: 2332 ms)
FAIL

FAIL (see details above)
----------------------------
Definitions kompiled: 2 (time: 0'7'' elapsed, 0'12'' CPU)
PDF posters kompiled: 2 (time: 0'10'' elapsed, 0'20'' CPU)
Programs krun: 16 (time: 0'6'' elapsed, 0'42'' CPU)
Total time: 0'23'' elapsed, 1'14'' CPU
----------------------------

fslwork$ /home/grosu/github.grosu/k/bin/krun '/home/grosu/github.grosu/k/tutorial/1_k/4_imp++/lesson_1/tests/io.imp' '--directory' 'imp-java_kompiled'
[6, 19.99 ms]
<T>
    <k>
        "Input two numbers: " ~> HOLE, .List{","} ~> print ( HOLE ) ; ~>
          int x, y ; ~> x = (read ( )) ; ~> y = (read ( )) ; ~> print (
          "Their sum is: ", (x + y), "\n" ) ;
    </k>
    <state>
        .Map
    </state>
</T>
fslwork$
```

Grigore

**Comment by yilongli** [comment]
That is the problem I am trying to fix. So it is probably because you didn't run `ant clean build`.

**Comment by grosu** [comment]
I definitely did `ant` and `ant clean`.  Here is the proof:

```
fslwork$ git status
# On branch unified-builtins2
nothing to commit, working directory clean
fslwork$ git log
commit e2082803b64107e2fab5bf599e165d5cacec2a88
Merge: d348389 0417ab7
Author: Yilong Li <yli147@illinois.edu>
Date:   Sat May 31 14:54:09 2014 -0500

    Merge pull request #515 from yilongli/refactoring-hashCode

    Refactoring hash code

commit 0417ab7d593a02cc2be4589165a6a492cd4a08fa
Author: Yilong Li <yli147@illinois.edu>
Date:   Sat May 31 13:45:41 2014 -0500

    addressed issues in code reviews

commit d3483891b79c187db530bc237c1b62d274be0ddf
Merge: b8285ff 8dd0560
Author: Andrei Stefanescu <andreistef@gmail.com>
Date:   Fri May 30 11:08:23 2014 -0500

    Merge pull request #517 from andreistefanescu/unified-builtins2

    enhanced builtin equality for symbolic terms; fix for immutable SymbolicCons

commit b8285ff515f31a2e3ed906b14da901998edcadd5
Merge: 718f396 5e61636
Author: dwightguth <dwight.guth@runtimeverification.com>
Date:   Thu May 29 16:32:48 2014 -0500

    Merge pull request #516 from dwightguth/floats

    add linux 32 bit mpfr

commit 5572b4b100b55a5104c968a3fb88a3b1a6d20836
Author: Yilong Li <yli147@illinois.edu>
Date:   Thu May 29 15:58:31 2014 -0500

    resolved new changes in the branch

commit f67c7ed1ff1f8e3afd996232c27b2ba69f27f29b
Author: Yilong Li <yli147@illinois.edu>
Date:   Thu May 29 15:24:46 2014 -0500

    refactored ConstrainedTerm to derive JavaSymbolicObject instead of Term

commit dc359536e4caa9eb5678a0360be5fc80f725c4aa
fslwork$ ant clean
Buildfile: /home/grosu/github.grosu/k/build.xml

clean:
   [delete] Deleting directory /home/grosu/github.grosu/k/src/javasources/parsers/Concrete/include
   [delete] Deleting directory /home/grosu/github.grosu/k/src/javasources/KTool/lib/resources/sdf

clean-java:
   [delete] Deleting directory /home/grosu/github.grosu/k/src/javasources/KTool/bin
   [delete] Deleting: /home/grosu/github.grosu/k/lib/java/k3.jar
   [delete] Deleting: /home/grosu/github.grosu/k/lib/version.txt

BUILD SUCCESSFUL
Total time: 0 seconds
fslwork$ ant
Buildfile: /home/grosu/github.grosu/k/build.xml

build:

concrete.signature:

concrete.signature.helper:
    [mkdir] Created dir: /home/grosu/github.grosu/k/src/javasources/parsers/Concrete/include
     [java]   including ../parsers/Concrete/syntax/Concrete.sdf
     [java]   including ../parsers/Concrete/syntax/Common.sdf
     [java]   including ../parsers/Concrete/syntax/KTechnique.sdf
     [java]   including ../parsers/Concrete/syntax/KBuiltinsBasic.sdf
     [java]   including ../parsers/Concrete/syntax/Integration.sdf
     [java]   including ../parsers/Concrete/syntax/Variables.sdf
     [echo] Packed to .def
     [java] [ org.strategoxt.tools.main-sdf2rtg | info ] Calling external tool /home/grosu/github.grosu/k/src/javasources/KTool/../../../lib/native/linux/sdf2table -i /home/grosu/github.grosu/k/src/javasources/KTool/../parsers/Concrete/include/Concrete.def -o /tmp/StrategoXT9125441388088478122.tmp -n -m Concrete
     [java] [ org.strategoxt.tools.main-sdf2rtg | info ] Calling external tool /home/grosu/github.grosu/k/src/javasources/KTool/../../../lib/native/linux/implodePT -i /tmp/StrategoXT9125441388088478122.tmp -o /tmp/StrategoXT8241356944012981185.tmp
     [echo] Def -> Rtg
     [echo] Build TBL

all.signature:

concrete.stratego2java:
     [copy] Copying 1 file to /home/grosu/github.grosu/k/src/javasources/KTool/lib/resources/sdf
     [copy] Copying 1 file to /home/grosu/github.grosu/k/src/javasources/KTool/lib/resources/sdf
     [copy] Copying 1 file to /home/grosu/github.grosu/k/src/javasources/KTool/lib/resources/sdf
     [copy] Copying 1 file to /home/grosu/github.grosu/k/src/javasources/KTool/lib/resources/sdf
     [copy] Copying 1 file to /home/grosu/github.grosu/k/src/javasources/KTool/lib/resources/sdf

concrete.stratego2java.helper:
     [java] [ Main | info ] Front-end succeeded         : [user/system] = [4.53s/0.00s]
     [java] [ Main | info ] Optimization succeeded -O 2 : [user/system] = [0.15s/0.00s]
     [java] [ Main | info ] Abstract syntax in 'ConcreteMain.rtree'
     [java] [ Main | info ] Export of externals succeeded : [user/system] = [0.05s/0.00s]
     [java] [ Main | info ] Backend preprocessing succeeded : [user/system] = [1.30s/0.00s]
     [java] [ Main | info ] Referenced imported term file include/Concrete.tbl; copying to ConcreteMain
     [java] [ Main | info ] Backend transformation succeeded : [user/system] = [0.52s/0.00s]
     [java] [ Main | info ] Backend optimization succeeded -O 2 : [user/system] = [0.61s/0.00s]
     [java] [ Main | info ] Backend postprocessing succeeded : [user/system] = [0.07s/0.00s]
     [java] [ Main | info ] Deleted existing java files in output directory ConcreteMain : [user/system] = [0.00s/0.00s]
     [java] [ Main | info ] Pretty printing succeeded   : [user/system] = [1.32s/0.00s]
     [java] [ Main | info ] Compilation succeeded       : [user/system] = [8.57s/0.00s]
     [copy] Copying 254 files to /home/grosu/github.grosu/k/src/javasources/KTool/src/org/kframework/parser/concrete/lib
   [delete] Deleting: /home/grosu/github.grosu/k/src/javasources/parsers/Concrete/trans/ConcreteMain.rtree
   [delete] Deleting: /home/grosu/github.grosu/k/src/javasources/parsers/Concrete/trans/ConcreteMain.dep
   [delete] Deleting directory /home/grosu/github.grosu/k/src/javasources/parsers/Concrete/trans/ConcreteMain

all.stratego2java:

build-subprojects:

init:
    [mkdir] Created dir: /home/grosu/github.grosu/k/src/javasources/KTool/bin
     [copy] Copying 30 files to /home/grosu/github.grosu/k/src/javasources/KTool/bin
     [copy] Copying 5 files to /home/grosu/github.grosu/k/src/javasources/KTool/bin
   [javacc] Java Compiler Compiler Version 5.0 (Parser Generator)
   [javacc] (type "javacc" with no arguments for help)
   [javacc] Reading from file /home/grosu/github.grosu/k/src/javasources/KTool/src/org/kframework/parser/basic/Basic.jj . . .
   [javacc] Note: UNICODE_INPUT option is specified. Please make sure you create the parser/lexer using a Reader with the correct character encoding.
   [javacc] File "TokenMgrError.java" does not exist.  Will create one.
   [javacc] File "ParseException.java" does not exist.  Will create one.
   [javacc] File "Token.java" does not exist.  Will create one.
   [javacc] File "SimpleCharStream.java" does not exist.  Will create one.
   [javacc] Parser generated successfully.
   [javacc] Java Compiler Compiler Version 5.0 (Parser Generator)
   [javacc] (type "javacc" with no arguments for help)
   [javacc] Reading from file /home/grosu/github.grosu/k/src/javasources/KTool/src/org/kframework/utils/kastparser/KastParser.jj . . .
   [javacc] Note: UNICODE_INPUT option is specified. Please make sure you create the parser/lexer using a Reader with the correct character encoding.
   [javacc] File "TokenMgrError.java" does not exist.  Will create one.
   [javacc] File "ParseException.java" does not exist.  Will create one.
   [javacc] File "Token.java" does not exist.  Will create one.
   [javacc] File "SimpleCharStream.java" does not exist.  Will create one.
   [javacc] Parser generated successfully.

build-project-java:
     [echo] KTool all: /home/grosu/github.grosu/k/src/javasources/KTool/build.xml
    [javac] Compiling 919 source files to /home/grosu/github.grosu/k/src/javasources/KTool/bin
    [javac] Note: Some input files use or override a deprecated API.
    [javac] Note: Recompile with -Xlint:deprecation for details.

build:

jar:
      [jar] Building jar: /home/grosu/github.grosu/k/lib/java/k3.jar

alllocal:

makeversion:
     [echo] ./lib/version.txt
     [echo] K-framework nightly build.
     [echo] Git Revision: d348389
     [echo] Build date: Mon Jun 2 22:08:10 CDT 2014

BUILD SUCCESSFUL
Total time: 27 seconds
fslwork$ cd tutorial/1_k/4_imp++/lesson_1/
fslwork$ ktest tests/config.xml 
Running initial scripts...
SUCCESS
Kompile the language definitions...(2 in total)
SUCCESS
Generate PDF files...(2 in total)
SUCCESS
Running /home/grosu/github.grosu/k/tutorial/1_k/4_imp++/lesson_1/imp.k programs... (8 in total, 0 with input, 8 with output, 0 with error)
Running /home/grosu/github.grosu/k/tutorial/1_k/4_imp++/lesson_1/imp.k programs... (8 in total, 0 with input, 8 with output, 0 with error)
ERROR: ['/home/grosu/github.grosu/k/bin/krun' '/home/grosu/github.grosu/k/tutorial/1_k/4_imp++/lesson_1/tests/io.imp' '--directory' 'imp-java_kompiled'] output doesn't match with expected output (time: 2243 ms)
ERROR: ['/home/grosu/github.grosu/k/bin/krun' '/home/grosu/github.grosu/k/tutorial/1_k/4_imp++/lesson_1/tests/spawn.imp' '--directory' 'imp-java_kompiled'] output doesn't match with expected output (time: 2294 ms)
ERROR: ['/home/grosu/github.grosu/k/bin/krun' '/home/grosu/github.grosu/k/tutorial/1_k/4_imp++/lesson_1/tests/sum-io.imp' '--directory' 'imp-java_kompiled'] output doesn't match with expected output (time: 2386 ms)
FAIL

FAIL (see details above)
----------------------------
Definitions kompiled: 2 (time: 0'10'' elapsed, 0'15'' CPU)
PDF posters kompiled: 2 (time: 0'10'' elapsed, 0'20'' CPU)
Programs krun: 16 (time: 0'6'' elapsed, 0'42'' CPU)
Total time: 0'26'' elapsed, 1'17'' CPU
----------------------------

fslwork$ /home/grosu/github.grosu/k/bin/krun '/home/grosu/github.grosu/k/tutorial/1_k/4_imp++/lesson_1/tests/io.imp' '--directory' 'imp-java_kompiled' --backend java
[6, 20.22 ms]
<T>
    <k>
        "Input two numbers: " ~> HOLE, .List{","} ~> print ( HOLE ) ; ~> 
          int x, y ; ~> x = (read ( )) ; ~> y = (read ( )) ; ~> print ( 
          "Their sum is: ", (x + y), "\n" ) ;
    </k>
    <state>
        .Map
    </state>
</T>
fslwork$ 
```

**Comment by yilongli** [comment]
The failure is really weird:http://fslwork.cs.illinois.edu/rv-jenkins/job/k-framework-pull-requests/506/testReport/junit/_samples_java_rewrite_engine_tutorial_1_k_4_imp++_imp/k/imp_k/

**Comment by yilongli** [comment]
Jenkins: test this please

**Comment by grosu** [comment]
OK, maybe it is me, tired.  I just pulled from the kframework/unified-builtins2 branch, but your updates are probably not merged into the branch yet, I guess.  How can I pull from your branch?

**Comment by yilongli** [comment]
@grosu Oh I see. These changes are still on my own fork `yilongli:unified-builtins2` (the pull request says I am trying to merge it to the `kframework:unified-builtins2`. So if you want to try out my patch, you will have to `fetch` my fork.

**Comment by yilongli** [comment]
@grosu I will paste the instructions here in a minute.

**Comment by yilongli** [comment]
@grosu 

```
git add remote yilongli https://github.com/yilongli/k
git fetch yilongli unified-builtins2
git checkout FETCH_HEAD
```

**Comment by grosu** [comment]
Great, lesson_1 works, thanks!  BTW, `git add remote` should be `git remote add`.  I prefer to do it this way:

```
git remote add yilongli https://github.com/yilongli/k
git pull --rebase yilongli:unified-builtins2
```

However, we have a problem with lesson_2:

```
fslwork$ pwd
/home/grosu/github.grosu/k/tutorial/1_k/4_imp++/lesson_2
fslwork$ ktest tests/config.xml 
Running initial scripts...
SUCCESS
Kompile the language definitions...(2 in total)
SUCCESS
Generate PDF files...(2 in total)
SUCCESS
Running /home/grosu/github.grosu/k/tutorial/1_k/4_imp++/lesson_2/imp.k programs... (8 in total, 0 with input, 8 with output, 0 with error)
ERROR: ['/home/grosu/github.grosu/k/bin/krun' '/home/grosu/github.grosu/k/tutorial/1_k/4_imp++/lesson_1/tests/div.imp'] output doesn't match with expected output (time: 2295 ms)
ERROR: ['/home/grosu/github.grosu/k/bin/krun' '/home/grosu/github.grosu/k/tutorial/1_k/4_imp++/lesson_1/tests/io.imp'] output doesn't match with expected output (time: 2399 ms)
ERROR: ['/home/grosu/github.grosu/k/bin/krun' '/home/grosu/github.grosu/k/tutorial/1_k/2_imp/lesson_1/tests/collatz.imp'] output doesn't match with expected output (time: 2436 ms)
ERROR: ['/home/grosu/github.grosu/k/bin/krun' '/home/grosu/github.grosu/k/tutorial/1_k/4_imp++/lesson_1/tests/locals.imp'] output doesn't match with expected output (time: 2435 ms)
ERROR: ['/home/grosu/github.grosu/k/bin/krun' '/home/grosu/github.grosu/k/tutorial/1_k/4_imp++/lesson_1/tests/sum-io.imp'] output doesn't match with expected output (time: 2472 ms)
ERROR: ['/home/grosu/github.grosu/k/bin/krun' '/home/grosu/github.grosu/k/tutorial/1_k/2_imp/lesson_1/tests/sum.imp'] output doesn't match with expected output (time: 2513 ms)
ERROR: ['/home/grosu/github.grosu/k/bin/krun' '/home/grosu/github.grosu/k/tutorial/1_k/4_imp++/lesson_1/tests/spawn.imp'] output doesn't match with expected output (time: 2511 ms)
ERROR: ['/home/grosu/github.grosu/k/bin/krun' '/home/grosu/github.grosu/k/tutorial/1_k/2_imp/lesson_1/tests/primes.imp'] output doesn't match with expected output (time: 2551 ms)
Running /home/grosu/github.grosu/k/tutorial/1_k/4_imp++/lesson_2/imp.k programs... (8 in total, 0 with input, 8 with output, 0 with error)
FAIL

FAIL (see details above)
----------------------------
Definitions kompiled: 2 (time: 0'6'' elapsed, 0'12'' CPU)
PDF posters kompiled: 2 (time: 0'10'' elapsed, 0'20'' CPU)
Programs krun: 16 (time: 0'6'' elapsed, 0'42'' CPU)
Total time: 0'22'' elapsed, 1'14'' CPU
----------------------------

fslwork$ /home/grosu/github.grosu/k/bin/krun '/home/grosu/github.grosu/k/tutorial/1_k/2_imp/lesson_1/tests/collatz.imp'
<T>
    <k>
        .K
    </k>
    <store>
        1 |-> 2
        2 |-> 1
        3 |-> 1
        4 |-> 3
        5 |-> 66
    </store>
    <env>
        m |-> 1
        n |-> 2
        q |-> 3
        r |-> 4
        s |-> 5
    </env>
</T>
fslwork$ more tests/collatz.imp.out 
<T>
    <k>
        .K
    </k>
    <env>
        m |-> 1
        n |-> 2
        q |-> 3
        r |-> 4
        s |-> 5
    </env>
    <store>
        1 |-> 2
        2 |-> 1
        3 |-> 1
        4 |-> 3
        5 |-> 66
    </store>
</T>
fslwork$ 
```

As you can see, the order of the env and store cells is different in the two backends.

**Comment by yilongli** [comment]
Oh, I see. This is annoying. But I will need some help here. If someone can tell me how the maude backend determines the order the cells, I can implement the same sorting criterion in the pretty printer.

**Comment by andreistefanescu** [comment]
I thought both backends use the same pretty printer, which sorts the cells
in the same way. But if it does not, we should change the pretty printer.

Andrei

**Comment by grosu** [comment]
But shouldn't Maude return configurations as ASTs, same like the java and any other backend, and then have the pretty printer print them based on the order of cells defined in the user-declared configuration?  I mean, why does this have anything to do with the backend?  I don't mean to criticize, just trying to understand.

Grigore

**Comment by yilongli** [comment]
Right, right, I misunderstood how the pretty printer works. I will fix the pretty printer to use the user-declared order in the configuration.

**Comment by yilongli** [comment]
I have updated the `*.out` files in all lessons, which should be double-checked of course. But there are two more issues:
1. There is no order specified for the output of multiplicity cells. We may need some way to specify in the configuration how to sort these multiplicity cells.
2. The ktest config files in lessons 3,6,7, and 8 are not working properly. Basically, there are no `imp-java_kompiled` directory generated after running ktest. In particular, running ktest for lesson 3 reports several errors, while running ktest on lessons 6,7, and 8 seems to kompile the defintions for the maude backend _twice_.

**Comment by traiansf** [comment]
I think the pretty printer actually uses that order (or at least, it used
to).  However, if multiplicity is involved, there is no clear way to sort
them.  One suggestion would be to generate the corresponding strings and
sort them alphabetically.  Of course we could add a tag "index" to the cell
we want to index on such as the <id> cell :-)

2014-06-03 7:06 GMT+03:00 Yilong Li notifications@github.com:

**Comment by grosu** [comment]
This is part of the canonization of configurations that we were talking about for so long.  Once we have this, then we can also test the equality of terms nicely.  Maybe the problem is trickier than it seems and maybe we should take it more seriously (i.e., to implement the algorithm for canonicality of graphs, which is used as part of the graph isomorphism algorithms/packages: http://en.wikipedia.org/wiki/Graph_canonization).

Grigore

**Comment by dwightguth** [comment]
We are currently sorting Maps, Lists, and Sets alphabetically, it makes sense to sort cells of multiplicity \* the same way.

File: src/javasources/KTool/src/org/kframework/backend/unparser/UnparserFilterNew.java [hunk 1]
```diff
@@ -257,6 +258,29 @@ public Void visit(Configuration configuration, Void _) {
...
     @Override
     public Void visit(Cell cell, Void _) {
+        Term contents = cell.getContents();
+        Bag sortedBag = null;
+        if (contents instanceof Bag) {
+            Bag origBag = (Bag) contents;
+            sortedBag = origBag.shallowCopy();
+            sortedBag.setContents(new ArrayList<Term>());
+            for (String cellLabel : context.getConfigurationStructureMap().get(cell.getLabel()).sons.keySet()) {
+                for (Term term : origBag.getContents()) {
```
Comment (dwightguth) [inline src/javasources/KTool/src/org/kframework/backend/unparser/UnparserFilterNew.java]: this is quadratic. Can't you create a Comparator and then call something like Collections.sort?

File: src/javasources/KTool/src/org/kframework/backend/unparser/UnparserFilterNew.java [hunk 2]
```diff
@@ -397,6 +425,14 @@ public Void visit(KApp kapp, Void _) {
                 } else{
                     int i = 0;
                     String [] rawLabelList = rawLabel.split("_");
+                    int lastIdx = termList.size() - 1;
+                    if (termList.get(lastIdx) instanceof ListTerminator) {
```
Comment (dwightguth) [inline src/javasources/KTool/src/org/kframework/backend/unparser/UnparserFilterNew.java:460]: this isn't at all the right solution, but this code is already a nasty mess, so I'm inclined to leave it for now.

**Comment by dwightguth** [comment]
This needs to add stuff to the regression test suite.

File: src/javasources/KTool/src/org/kframework/backend/unparser/UnparserFilterNew.java (hunk 1 above)
Comment (yilongli) [inline src/javasources/KTool/src/org/kframework/backend/unparser/UnparserFilterNew.java]: Yes, quadratic, but how many cells do you think can appear in a bag? Besides, there are terms other than cells in the bag, I don't see any benefit to write more complicated code for this.

Comment (dwightguth) [inline src/javasources/KTool/src/org/kframework/backend/unparser/UnparserFilterNew.java]: Okay, take a look at the code I wrote in ConcretizeSyntax for SetBuiltin and MapBuiltin. You should do something similar there, while keeping the unparser simple.

Comment (yilongli) [inline src/javasources/KTool/src/org/kframework/backend/unparser/UnparserFilterNew.java]: OK, your code makes sense. I will refactor my code.

**Comment by yilongli** [comment]
Once all the lessons are (partially) fixed, they will be included in the regression test suite before merge.

**Comment by grosu** [comment]
I am not sure that alphabetical sorting the right way to go, but so be it for now, we can get back to it later.  What I don't like is that maps are now pretty printed as

```
1 |-> a
10 |-> j
2 |> b
3 |-> c
...
```

But anything else would be more complicated and currently not the effort, so let's go with this.  I think this is understood, but just to be sure we are all on the same page: so the idea is to first sort cells according to the user-defined cell order in the configuration, including inside multiplicity \* cells, and in the case of multiplicity \* to further sort the various clones in alphabetical order by the first subcell, then by the second, etc., right?  This also gives us a configuration canonization, doesn't it?

**Comment by yilongli** [comment]
@grosu Yes, this is the criterion I used in the pretty printer. As for the order of elements in the map, if you want, I can sort the elements first by length then by alphabetical order. It's a pretty simple fix.

**Comment by dwightguth** [comment]
I'm not sure that's the right solution unfortunately because then you would
have:

at
pot
bike
apple

etc.

The way linux does this in its "sort" utility is via "sort -n" which treats
the values being sorted as numeric. Technically I think the best you can do
is check whether the key of the map is IntBuiltin, FloatBuiltin, etc and if
so sort by compareTo on the BigInteger/BigFloat.

**Comment by andreistefanescu** [comment]
We want to have something similar to what "sort -n" does, which also sort
a1, a2, a10 in this order. I think we can use existing Java implementations
(take a look here
http://stackoverflow.com/questions/1262239/natural-sort-order-string-comparison-in-java-is-one-built-in).
It should not be very complicated.

Andrei

**Comment by dwightguth** [comment]
go ahead and put it under src/javasources/KTool/src/com/davekoelle, unless anyone has some strong objection to it being placed there.

**Comment by dwightguth** [comment]
http://www.davekoelle.com/alphanum.html I mean

**Comment by yilongli** [comment]
@grosu @andreistefanescu The `superheat` and `supercool` do not work in the Java backend, and it's not a trivial fix. I have looked into the four kompilation passes that are related, that is, `AddSupercoolDefinition`, `AddHeatingConditions`, `AddSuperheatRules`, and `ResolveSupercool`; and I believe the problem is the following rules generated in the `AddHeatingConditions` pass:

```
    syntax KLabel ::= "heat"
                    | "heated"
                    | "redex"
                    | "cool"

    rule heat(heated(K:K)) => K

    rule heat(heated(_:KList,, K:K,, _)) => K [heat-choice]

    rule heat((redex(_:K,, K:K,, KL:KList) => K,, KL),, _) [owise]

    rule heat((redex(K:K) => .KList),, _:KList,, heated(_,, (.KList => K))) [
      owise]

    rule cool(K:K) => K [owise]
```

Fixing it requires at least associative matching on `KList`.

Other passes seem to generate pretty normal rules and they shouldn't cause many problems.

**Comment by yilongli** [comment]
@dwightguth I need your help to fix the checkstyle config file to exclude the `AlphanumComparator.java` file.

@grosu Right now, this pull request is blocked by the issue mentioned in https://github.com/kframework/k/issues/545. Since `superheat` and `supercool` will not be supported by the Java backend very soon. Those broken ktest config files in lesson 3, 6, 7, and 8 need to be modified; however, I am not familiar with ktest.

**Comment by dwightguth** [comment]
Modify k/build.xml and add the following line after line 206:

`<exclude name="src/com/**" />`

**Comment by dwightguth** [comment]
actually make that src/com/davekoelle just to be on the safe side, in case we decide we want to write code in com.runtimeverification or something.

**Comment by dwightguth** [comment]
@yilongli  can you also please update k/LICENSE to mention the new source code and its license under the LGPL?

**Comment by yilongli** [comment]
I have tried to exclude `<exclude name="src/com/davekoelle/**" />` in checkstyle config file, but it's still failing. Besides, it works on my machine. I am not sure what's wrong.

**Comment by yilongli** [comment]
I would like to get this pull request merge asap.

@dwightguth please review again. I wasn't able to exclude the Alphanum source files from checkstyle copyright check following your instruction. So I tried to modify the build.xml differently, but I am not sure this is the correct way.

@grosu please review the updated *.out files in case you haven't done that.

**Comment by grosu** [comment]
I agree that we should get this pull request merged asap!

In fact, I think we should get all these pull requests related to updated languages in unified-builtins2 merged asap, because they will help us better test everything else we do.

@dwightguth let us please give these priority
@yilongli I'm pretty sure I checked the .out files.  But as I said in a previous message, it would be good to not worry at all about the java backend before we do the merge; so I would remove all the tests related to the java backend from the config.xml files.  We need to have a discussion about how to modularly add new backends to ktest.

**Comment by yilongli** [comment]
Yeah, all tests of imp/imp++ that are related to the Java backend are removed.

**Comment by dwightguth** [comment]
This is still not actually testing imp++ as part of Jenkins. However, if you uncomment it and the tests pass, you can merge this. If they don't, I'd like to take a quick look at what is needed to fix them before merging.

**Comment by yilongli** [comment]
**UPDATE**: The output is actually correct. The problem is that I forgot to use natural sorting on cells as well.

The only thing that failed is due to the new sorting algorithm used in map. I will update the output and merge this.

**Pull request merged**